python/toolchain.bzl

```diff
@@ -0,0 +1,21 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""All toolchains re-exported here"""
```

Comment:
I thought it would be good to evolve the repo towards a more standardised structure like the one described by [rules-template](https://github.com/bazel-contrib/rules-template)

Alternatively, if we would prefer to paint that bikeshed another time, I can hide as much of the implementation inside //python/private/pypi and we can restructure later.

Reply:
I don't have strong opinions about the layout of everything under //python/private, since its all private, implementation details of how we structure our code.

What I do care about is avoiding "kitchen sink" public interfaces. The issue is that Bazel has to evaluate the transitive closure of loads() regardless of what is actually used. This can cause oddities when code is evaluated in a different phase (notably, if the repo-phase loads loading-phase code, the Starlark environment isn't quite the same).

More notably, the unused loads work against loading-phase caching. As an example this toolchain.bzl demonstrates: if someone is only using UvToolchainInfo, then they don't need to load the uv_toolchain implementation, and changes to that implementation shouldn't invalidate the loading-phase. However, by having the two exported in the same file, changes to one will always invalidate the other. This problem is further compounded because it's a toilsome state to get out of because, so long as any of your transitive closure are using the kitchen sink targets, the whole build pays the cost. So you have to update many things to realize the benefits. This is why you see e.g. `//python:py_binary.bzl` et al.

So for a public API, please create separate files to re-export the provider and rule. That at least makes separating them in the future to realize the benefits just an internal refactoring, and thus a much more tenable change.

Reply:
No objections from me. Will FIXUP.